### PR TITLE
Enhance the fold type to capture terminating folds

### PR DESCRIPTION
**Pull request by adithyaov**
See #770 and #775 

The changes made are minimally required.
Only in a few places, I've corrected the variables.
In places where the entire function was refactored, I've taken the liberty of variable naming and indentation, completely different from the original. (I'll go over and reduce the diff)

### CODE REVIEW

File: src/Streamly/Internal/Data/Fold.hs [hunk 1]
```diff
@@ -1169,19 +1320,35 @@ distribute = foldr foldCons foldNil
...
+
     where
-    initial    = Prelude.mapM (\(Fold s i e) ->
-        i >>= \r -> return (Fold s (return r) e)) fs
-    step ss a  = do
-        Prelude.mapM_ (\(Fold s i _) -> i >>= \r -> void (s r a)) ss
-        return ss
-    extract =
-        Prelude.mapM_ (\(Fold _ i e) -> i >>= \r -> e r)
+
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold.hs:1325]: This is pretty ugly. Improve code.

File: src/Streamly/Internal/Data/Fold.hs [hunk 2]
```diff
@@ -1226,18 +1393,63 @@ distribute_ fs = Fold step initial extract
...
+        sL <- beginL
+        sR <- beginR
+        return $ RunBoth sL sR
+
+    step (RunBoth sL sR) a = do
+        r <- f a
+        case r of
+            Left b -> do
+                res <- stepL sL b
+                return
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold.hs:1410]: Rethink indentation, Unnecessary wastage of horizontal space.

File: src/Streamly/Internal/Data/Fold.hs [hunk 3]
```diff
@@ -1316,31 +1528,33 @@ demuxWith f kv = Fold step initial extract
...
+    step (Tuple' bmp mp) a = do
         let (k, a') = f a
-        in case Map.lookup k mp of
-            Nothing -> return mp
-            Just (Fold step' acc extract') -> do
-                !r <- acc >>= \x -> step' x a'
-                return $ Map.insert k (Fold step' (return r) extract') mp
-#endif
-    extract = Prelude.mapM (\(Fold _ acc e) -> acc >>= e)
+        case Map.lookup k mp of
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold.hs:1537]: This is pretty ugly. Improve code.

File: src/Streamly/Internal/Data/Fold.hs [hunk 4]
```diff
@@ -1367,23 +1583,68 @@ demuxWithDefault_ f kv (Fold dstep dinitial dextract) =
...
-            Nothing -> do
-                acc <- dstep dacc (k, a')
-                return (Tuple' mp acc)
-            Just (Fold step' acc _) -> do
-                _ <- acc >>= \x -> step' x a'
-                return (Tuple' mp dacc)
-    extract (Tuple' mp dacc) = do
+        ini <- dinitial
+        mp <- Prelude.mapM initialize kv
+        return $ DemuxingWithDefault ini Set.empty mp
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold.hs:1589]: This is pretty ugly. Improve code.

File: src/Streamly/Internal/Data/Fold.hs [hunk 5]
```diff
@@ -1411,20 +1672,29 @@ demuxWith_ f kv = Fold step initial extract
...
+    step (Tuple' n mp) a = do
+        let (k, a') = f a
+        case Map.lookup k mp of
+            Nothing -> return $ Partial $ Tuple' n mp
+            Just (Fold stp ini dn) -> do
+                !st <- ini
+                !res <- stp st a'
+                let n1 = n - 1
+                return
+                    $ case res of
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold.hs:1686]: Rethink indentation

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 6]
```diff
@@ -441,9 +539,10 @@ ltakeWhile predicate (Fold step initial done) = Fold step' initial' done'
 --
 -- @since 0.7.0
 {-# INLINABLE duplicate #-}
-duplicate :: Applicative m => Fold m a b -> Fold m a (Fold m a b)
-duplicate (Fold step begin done) =
-    Fold step begin (\x -> pure (Fold step (pure x) done))
+duplicate ::
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:542]: I don't think there is a way to write `duplicate` in a nice way.

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 7]
```diff
@@ -537,55 +669,79 @@ lchunksOf2 n (Fold step1 initial1 extract1) (Fold2 step2 inject2 extract2) =
 -- @
 {-# INLINE lsessionsOf #-}
 lsessionsOf :: MonadAsync m => Double -> Fold m a b -> Fold m b c -> Fold m a c
-lsessionsOf n (Fold step1 initial1 extract1) (Fold step2 initial2 extract2) =
-    Fold step' initial' extract'
+lsessionsOf n (Fold stepS initialS extractS) (Fold stepC initialC extractC) =
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Fold/Types.hs:672]: This code has been completely changed. The extension of the previous implementation is very buggy.

Read `XXX` above.

File: src/Streamly/Internal/Data/Stream/StreamD/Type.hs [hunk 8]
```diff
@@ -64,8 +64,10 @@ import Fusion.Plugin.Types (Fuse(..))
 import Streamly.Internal.Data.SVar (State(..), adaptState, defState)
 import Streamly.Internal.Data.Fold.Types (Fold(..), Fold2(..))
 
+import qualified Streamly.Internal.Data.Fold.Types as FL
 import qualified Streamly.Internal.Data.Stream.StreamK as K
 
+
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Stream/StreamD/Type.hs:70]: extra line

File: src/Streamly/Internal/Data/Parser/ParserD/Types.hs [hunk 9]
```diff
@@ -457,8 +459,10 @@ splitMany (Fold fstep finitial fextract) (Parser step1 initial1 extract1) =
                 return $ Continue n (Tuple3' s (cnt1 - n) fs)
             Done n b -> do
                 s <- initial1
-                fs1 <- fstep fs b
-                return $ Partial n (Tuple3' s 0 fs1)
+                sfs <- fstep fs b
+                case sfs of
+                    FL.Partial fs1 -> return $ Partial n (Tuple3' s 0 fs1)
+                    FL.Done fb -> return $ Done n fb
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser/ParserD/Types.hs:465]: Factor out `return`.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
Submitting partial review.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: src/Streamly/Internal/Data/Fold.hs [hunk 10]
```diff
@@ -965,23 +1056,46 @@ spanBy cmp (Fold stepL initialL extractL) (Fold stepR initialR extractR) =
     Fold step initial extract
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply src/Streamly/Internal/Data/Fold.hs:1056]: Why? The implementation and behaviour is correct. We might as well keep it.

File: src/Streamly/Internal/Data/Fold.hs [hunk 11]
```diff
@@ -1367,23 +1583,68 @@ demuxWithDefault_ f kv (Fold dstep dinitial dextract) =
...
+                return
+                    $ case res of
+                          Partial sres ->
+                              let fld = (Fold stp (return sres) dn)
+                                  mp1 = Map.insert k fld mp
+                               in Partial $ DemuxingWithDefault dacc bst mp1
+                          Done _ ->
+                              if Map.size mp == 1
+                              then Done ()
+                              else let bst1 = Set.insert k bst
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply src/Streamly/Internal/Data/Fold.hs:1618]: We will probably need a state monad or an IORef to check the termination condition.

[the remark this answers is not in the document]
Comment (adithyaov) [reply src/Streamly/Internal/Data/Fold.hs:1618]: I'll change this after 671

File: src/Streamly/Internal/Data/Fold/Types.hs [hunk 12]
```diff
@@ -460,8 +559,13 @@ initialize (Fold step initial extract) = do
...
     i <- initial
-    r <- step i a
-    return $ Fold step (return r) extract
+    res <- step i a
+    return $ case res of
+        Partial s -> Fold step (return s) extract
+        Done b ->
+            let step1 _ _ = return $ Done b
+                extract1 _ = return b
+             in Fold step1 (return i) extract1
```
[the remark this answers is not in the document]
Comment (adithyaov) [reply src/Streamly/Internal/Data/Fold/Types.hs:568]: This is a bug.

File: src/Streamly/Internal/Data/Fold.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (adithyaov) [reply src/Streamly/Internal/Data/Fold.hs:1537]: We will probably need a state monad or an IORef to check the termination condition if we use alterF.

**Comment by adithyaov** [comment]
See #770

**Pull request closed without merging**